src/Microsoft.ML.TimeSeries/SsaSpikeDetector.cs

```diff
@@ -133,6 +134,15 @@ private static IDataTransform Create(IHostEnvironment env, ModelLoadContext ctx,
             return new SsaSpikeDetector(env, ctx).MakeDataTransform(input);
         }
 
+        public override IStatefulTransformer Clone()
+        {
+            var clone = (SsaSpikeDetector)MemberwiseClone();
+            clone.Model = clone.Model.Clone();
+            clone.StateRef = (State)clone.StateRef.Clone();
```

Comment:
Should this be atomic operation, so Model and StateRef are consistent with each other? Could it happen that after Model is cloned but StateRef is not yet, the original StateRef changes, so then cloning StateRef would break consistency of Model & StateRef couple?

Reply:
Also looks like we have 1 to 1 correspondence between Model and State. maybe you could make a State a part of a Model now? So State object is tracked inside the Model 

---
In reply to: [236522741](https://github.com/dotnet/machinelearning/pull/1727#discussion_r236522741) [](ancestors = 236522741)

Reply:
I don't think so, since cloning happens when a prediction engine is created and it clones from a transformer that will not mutate it's state. The prediction engine keeps a copy and it mutates it and nobody clones the transformer reference maintained by the prediction engine because I don't allow access to it. 

---
In reply to: [236522741](https://github.com/dotnet/machinelearning/pull/1727#discussion_r236522741) [](ancestors = 236522741)